### PR TITLE
Contacts favourites grid view

### PR DESCRIPTION
**Pull request by Merkost**

### CODE REVIEW

File: app/src/main/res/menu/menu.xml [hunk 1]
```diff
@@ -18,6 +18,10 @@
         android:icon="@drawable/ic_dialpad_vector"
         android:title="@string/dialpad"
         app:showAsAction="ifRoom" />
+    <item
+        android:id="@+id/change_view_type"
+        android:title="@string/change_view_type"
+        app:showAsAction="never" />
```
Comment (tibbi) [inline app/src/main/res/menu/menu.xml:24]: why Never? We have an icon for this action.

Comment (Merkost) [reply app/src/main/res/menu/menu.xml:24]: Oh, right. I'll fix that in diaper too. Thanks.

**Comment by tibbi** [comment]
update to latest commons please

File: app/src/main/res/menu/menu.xml [hunk 2]
```diff
@@ -18,6 +18,17 @@
...
         app:showAsAction="ifRoom" />
+    <item
+        android:id="@+id/change_view_type"
+        android:icon="@drawable/ic_change_view_vector"
+        android:title="@string/change_view_type"
+        app:showAsAction="ifRoom" />
+    <item
+        android:id="@+id/column_count"
+        android:icon="@drawable/ic_column_count_vector"
+        android:showAsAction="never"
```
Comment (tibbi) [inline app/src/main/res/menu/menu.xml:29]: why never?

Comment (tibbi) [inline app/src/main/res/menu/menu.xml:29]: why never, and ifRoom under?

**Comment by tibbi** [comment]
refresh top menu on changing view type so the column count is properly refreshed

File: app/src/main/res/menu/menu.xml (hunk 2 above)
Comment (Merkost) [reply app/src/main/res/menu/menu.xml:29]: Oh, that's a typo, thank you.

**Comment by tibbi** [comment]
reordering favorites with drag & drop is actually broken in grid view, the rest seems ok

**Comment by Merkost** [comment]
> reordering favorites with drag & drop is actually broken in grid view, the rest seems ok

Yep, implemented drag & drop for gris now, thanks

**Comment by tibbi** [comment]
if i change the column count, theres is basically endless scrolling now..

**Comment by tibbi** [comment]
or no, it is not endless scrolling... It just shows all contacts instead of favorites only.

**Comment by Merkost** [comment]
> or no, it is not endless scrolling... It just shows all contacts instead of favorites only.

had to create another variable for favourite contacts and use it in updateFavoritesColumnCount function

File: app/src/main/kotlin/com/simplemobiletools/contacts/pro/fragments/MyViewPagerFragment.kt [hunk 3]
```diff
@@ -157,34 +153,48 @@ abstract class MyViewPagerFragment(context: Context, attributeSet: AttributeSet)
         }
     }
 
-    private fun sortByCustomOrder(starred: List<Contact>): ArrayList<Contact> {
-        val favoritesOrder = activity!!.config.favoritesContactsOrder
+    private fun List<Contact>.sortFavourites(shouldSort: Boolean = true): List<Contact> {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/contacts/pro/fragments/MyViewPagerFragment.kt:156]: this looks like a messed up function, theres too many returns

File: app/src/main/res/menu/menu.xml [hunk 4]
```diff
@@ -18,14 +18,22 @@
...
+        android:title="@string/column_count"
+        app:showAsAction="ifRoom" />
     <item
         android:id="@+id/import_contacts"
-        android:showAsAction="never"
         android:title="@string/import_contacts_from_vcf"
         app:showAsAction="never" />
     <item
         android:id="@+id/export_contacts"
-        android:showAsAction="never"
```
Comment (tibbi) [inline app/src/main/res/menu/menu.xml:28]: lets keep the nevers shown twice to fix some glitches.. Ifrooms dont need it.

File: app/src/main/kotlin/com/simplemobiletools/contacts/pro/fragments/MyViewPagerFragment.kt (hunk 3 above)
Comment (Merkost) [reply app/src/main/kotlin/com/simplemobiletools/contacts/pro/fragments/MyViewPagerFragment.kt:156]: I've improved the readability of the function, now it's clear what it does. Could you check it now please?

File: app/src/main/kotlin/com/simplemobiletools/contacts/pro/activities/MainActivity.kt [hunk 5]
```diff
@@ -233,6 +240,28 @@ class MainActivity : SimpleActivity(), RefreshContactsListener {
...
 
+    private fun changeViewType() {
+        ChangeViewTypeDialog(this) {
+            refreshMenuItems()
+            favorites_fragment?.updateFavouritesAdapter()
+        }
+    }
+
+    private fun changeColumnCount() {
+        val items = (CONTACTS_GRID_MIN_COLUMNS_COUNT..CONTACTS_GRID_MAX_COLUMNS_COUNT).map {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/contacts/pro/activities/MainActivity.kt:251]: man, stop overcomplicating things... Just take the solution from other apps for consistency, do not create your own solutions for everything. Minimal value is 1 so keep it simple. Just check how are such arraylists created in Simple App Launcher or other apps using it.

Comment (tibbi) [reply app/src/main/kotlin/com/simplemobiletools/contacts/pro/activities/MainActivity.kt:251]: whole CONTACTS_GRID_MIN_COLUMNS_COUNT isnt even needed, it should be removed from commons

**Comment by Merkost** [comment]
Deleted CONTACTS_GRID_MIN_COLUMNS_COUNT  constant from commons: https://github.com/SimpleMobileTools/Simple-Commons/pull/1758

File: app/src/main/kotlin/com/simplemobiletools/contacts/pro/adapters/ContactsAdapter.kt [hunk 6]
```diff
@@ -471,4 +491,26 @@ class ContactsAdapter(
     override fun onRowClear(myViewHolder: ViewHolder?) {
         onDragEndListener?.invoke()
     }
+
+    override fun zoomIn() {
+        val layoutManager = recyclerView.layoutManager
+        if (layoutManager is GridLayoutManager) {
+            val currentSpanCount = layoutManager.spanCount
+            val newSpanCount = (currentSpanCount - 1).coerceIn(CONTACTS_GRID_MIN_COLUMNS_COUNT, CONTACTS_GRID_MAX_COLUMNS_COUNT)
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/contacts/pro/adapters/ContactsAdapter.kt:499]: you just removed this constant in another PR...

Comment (Merkost) [reply app/src/main/kotlin/com/simplemobiletools/contacts/pro/adapters/ContactsAdapter.kt]: I was asking you about this at the moment, sorry.

File: app/src/main/kotlin/com/simplemobiletools/contacts/pro/adapters/ContactsAdapter.kt [hunk 7]
```diff
@@ -471,4 +491,26 @@ class ContactsAdapter(
     override fun onRowClear(myViewHolder: ViewHolder?) {
         onDragEndListener?.invoke()
     }
+
+    override fun zoomIn() {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/contacts/pro/adapters/ContactsAdapter.kt:495]: zooming should be handled by the fragment or activity, just like in other apps..

Comment (tibbi) [reply app/src/main/kotlin/com/simplemobiletools/contacts/pro/adapters/ContactsAdapter.kt:495]: and do the max and min column count checks consistently with other apps too

File: app/src/main/kotlin/com/simplemobiletools/contacts/pro/fragments/MyViewPagerFragment.kt [hunk 8]
```diff
@@ -157,34 +153,45 @@ abstract class MyViewPagerFragment(context: Context, attributeSet: AttributeSet)
         }
     }
 
-    private fun sortByCustomOrder(starred: List<Contact>): ArrayList<Contact> {
-        val favoritesOrder = activity!!.config.favoritesContactsOrder
-
-        if (favoritesOrder.isEmpty()) {
-            return ArrayList(starred)
+    private fun List<Contact>.sortFavourites(shouldSort: Boolean = true): List<Contact> {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/contacts/pro/fragments/MyViewPagerFragment.kt:156]: whats that? A private list extension?... Why not just a private function?

Comment (Merkost) [reply app/src/main/kotlin/com/simplemobiletools/contacts/pro/fragments/MyViewPagerFragment.kt:156]: By making it an extension function, it makes the code more convenient to use and could improve readability and also it encapsulates the sorting logic within the extension function and keeps it closely related to the Contact class. No worries, I'll replace it with a private function if you prefer it that way more

Comment (tibbi) [reply app/src/main/kotlin/com/simplemobiletools/contacts/pro/fragments/MyViewPagerFragment.kt]: of course, extensions shouldnt be hidden in such files, that kills their point

**Comment by tibbi** [comment]
ok thanks, should be good now

**Pull request merged**